strawberryfields/program.py

```diff
@@ -458,20 +435,39 @@ def append(self, op, reg):
         self.circuit.append(Command(op, reg))
         return reg
 
+    def _linked_copy(self):
+        """Create a copy of the Program, linked to the original.
+
+        Both the original and the copy are :meth:`locked <lock>`, since they share their RegRefs.
+
+        Returns:
+            Program: a copy of the Program
+        """
+        self.lock()
+        p = copy.copy(self)  # shares RegRefs with the source
+        # link to the original source Program
+        if self.source is None:
+            p.source = self
+        else:
+            p.source = self.source
+        return p
 
-    def compile(self, backend='fock', **kwargs):
-        """Compile the program for the given backend.
+    def compile(self, target, **kwargs):
+        """Compile the program targeting the given circuit template.
 
-        The compilation step validates the program, making sure all the Operations
-        used are accepted by the target backend.
-        Additionally it may decompose certain gates into sequences of simpler gates.
+        Validates the program against the given target, making sure all the Operations
+        used are accepted by the target template.
+        Additionally, depending on the target, the compilation may modify the quantum circuit
+        into an equivalent circuit, e.g. by decomposing certain gates into sequences
+        of simpler gates, or optimizing the gate ordering using commutation rules.
 
-        The compiled program shares its RegRefs with the original, which makes it easier
-        to access the measurement results, but also necessitates the locking of both the
-        compiled program and the original to make sure the RegRef state remains consistent.
+        The returned compiled Program shares its :class:`RegRefs <RegRef>` with the original,
+        which makes it easier to access the measurement results, but also necessitates the
+        :meth:`locking <lock>` of both the compiled program and the original to make sure the
```

Comment:
Would locking prevent a circuit from being compiled twice, since the RegRefs will be locked on second compile? 

Use-case: user compiles for a target, then engine compiles for a backend.

Reply:
No, `Program.lock()` just prevents the appending of additional Commands to the circuit, and changing the RegRef state. Compiling at the moment does not affect the RegRefs (although in future I suppose it could, for example eliminating unnecessary/unused modes?)